File: cli/node/src/cli/commands/memory/update.ts
import { _validateExpires } from '../../../application/memory/expiration.js';
import { parseObject } from '../../../application/memory/input.js';
import type { JsonObject } from '../../../backend/json.js';
import type { Backend } from '../../../backend/types.js';
import { printError, printSuccess, timedStatus } from '../../../output/branding.js';
import { formatAgentEnvelope, formatJson } from '../../../output/format.js';
import { exit } from '../../../runtime/exit.js';
import { setCurrentCommand } from '../../../runtime/state.js';

/**
 * 校验更新字段后修改指定记忆。
 * @param backend - 已装配的后端连接器。
 * @param memoryId - 目标记忆 ID。
 * @param text - 待解析或提交的文本。
 * @param opts - 本次操作选项。
 * @param opts.metadata - 记忆元数据的 JSON 文本。
 * @param opts.expires - 记忆到期日期，格式为 YYYY-MM-DD。
 * @param opts.timestamp - 业务时间戳。
 * @param opts.output - 终端输出格式。
 */
export async function cmdUpdate(
  backend: Backend,
  memoryId: string,
  text: string | undefined,
  opts: {
    metadata?: string;
    expires?: string;
    timestamp?: number;
    output: string;
  }
): Promise<void> {
  setCurrentCommand('update');
  let meta: JsonObject | undefined;
  if (opts.metadata) {
    try {
      meta = parseObject(opts.metadata);
    } catch {
      printError('Invalid JSON in --metadata.');
      exit(1);
    }
  }

  if (opts.expires) _validateExpires(opts.expires);

  const start = performance.now();
  let result: JsonObject;
  try {
    result = await timedStatus('Updating memory...', async () => {
      return backend.update(memoryId, text, meta, {
        expirationDate: opts.expires,
        timestamp: opts.timestamp,
      });
    });
  } catch (e) {
    printError(e instanceof Error ? e.message : String(e));
    exit(1);
  }
  const elapsed = (performance.now() - start) / 1000;

  if (opts.output === 'agent') {
    formatAgentEnvelope({
      command: 'update',
      data: result,
      durationMs: Math.round(elapsed * 1000),
    });
  } else if (opts.output === 'json') {
    formatJson(result);
  } else if (opts.output !== 'quiet') {
    printSuccess(`Memory ${memoryId.slice(0, 8)} updated (${elapsed.toFixed(2)}s)`);
  }
}

File: cli/python/src/memgo_cli/cli/commands/memory/update.py
"""记忆 update 用例。"""

from __future__ import annotations

import json
import time as _time

import typer

from memgo_cli.backend.json import json_object
from rich.console import Console

from memgo_cli.application.memory.expiration import _validate_expires
from memgo_cli.backend.types import Backend
from memgo_cli.output.branding import (
    print_error,
    print_success,
    timed_status,
)
from memgo_cli.output.format import (
    format_agent_envelope,
    format_json,
)

console = Console()
err_console = Console(stderr=True)


def cmd_update(
    backend: Backend,
    memory_id: str,
    text: str | None,
    *,
    metadata: str | None,
    expires: str | None,
    timestamp: int | None,
    output: str,
) -> None:
    """更新指定记忆的内容或元数据。"""
    from memgo_cli.runtime.state import is_agent_mode, set_current_command

    set_current_command("update")
    if is_agent_mode():
        output = "agent"
    meta = None
    if metadata:
        try:
            meta = json_object(json.loads(metadata))
        except ValueError:
            print_error(err_console, "Invalid JSON in --metadata.", hint=None)
            raise typer.Exit(1) from None

    if expires:
        _validate_expires(expires)

    _start = _time.perf_counter()
    with timed_status(err_console, "Updating memory...") as _ts:
        try:
            result = backend.update(
                memory_id,
                content=text,
                metadata=meta,
                expiration_date=expires,
                timestamp=timestamp,
            )
        except Exception as e:
            print_error(err_console, str(e), hint=None)
            raise typer.Exit(1) from None
    _elapsed = _time.perf_counter() - _start

    if output == "agent":
        format_agent_envelope(
            console,
            command="update",
            data=result,
            duration_ms=int(_elapsed * 1000),
            scope=None,
            count=None,
        )
    elif output == "json":
        format_json(console, result)
    elif output != "quiet":
        print_success(console, f"Memory {memory_id[:8]} updated ({_elapsed:.2f}s)")
